### DISCUSSION TITLE
Improving Installation Docs / Distribution Methods for Debian

### DISCUSSION THREAD

**Discussion by jeffgeorge** [sabnzbd/sabnzbd#2706] [Ideas]
Over the past couple weeks/months -- particularly since Debian 12 was released in June -- we've periodically had people drop by Discord with a story along these lines:

* I just installed SABnzbd! On Debian!
* But It's SO OLD!
* How do I get it to be Less Old??!?!
* BUT NOT DOCKER! **NO DOCKER!** DO **NOT** SUGGEST DOCKER!

In my assessment, your average user asking for help installing SABnzbd on Debian _probably_ isn't ready to drop what they're doing and learn how to run a python app from scratch, with a venv, and figure out how to run it as a service manually, and keep it up to date, if all they're used to is installing stuff via `apt`.

So . . how do we help these users?

I know "Just Update The Official Apt Repo Faster" is off the table. I also know "Use the Ubuntu PPA" is apparently off the table. So what _are_ our options?

Is it completely insane to set up our own deb build pipeline through Github Actions? Would it be an immense mess to set up our own apt repo? Should we be directing people to snaps instead? flatpak? I don't know what the best approach is going to be because _my_ approach -- as someone who has been maintaining a handful of personal Debian servers for 20+ years -- has been "if you care about bleeding-edge application versions, use Docker"... so this hasn't been a problem I've had to personally solve.

Pinging @jcfp as our resident Debian/Ubuntu Maintainer.

**Comment by bakerboy448** [comment]
> your average user asking for help installing SABnzbd on Debian probably isn't ready to drop what they're doing and learn how to run a python app from scratch, with a venv, and figure out how to run it as a service manually, and keep it up to date

provide an install/update script?

see qbm and pmm docs for similar ideas accordingly

**Reply by jeffgeorge** [reply]
Can you point me in a specific direction re: qbm / pmm? The only pmm install script I was able to find was for [Percona Monitoring and Management](https://github.com/percona/pmm/blob/main/get-pmm.sh) . . . which largely just installs Docker for you.

**Reply by bakerboy448** [reply]
https://metamanager.wiki/en/latest/home/guides/local.html

Plex Meta Manager

**Reply by jeffgeorge** [reply]
Ok, what I see there is an incredibly verbose means of installing in a venv, no scripting? This is a great model for getting people to set up a venv, but I'd prefer we figure out some better way of packaging and distributing for Debian that doesn't expect people to _do_ all that.

**Comment by sanderjo** [comment]
Did you see https://github.com/sabnzbd/sabnzbd/discussions/2594 ?

**Reply by sanderjo** [reply]
If you provide a package / distribution method for Debian ... how about Redhat/CentOS, SuSE, Arch, and the dozens other Linux distro's?

I see the follow possibilities for that, i.e. distro indepent solutions:
* automatic python-venv script that always works on (about) any Linux
* indeed snap / flatplak. Linux purists hate it, but hey, linux purists are not the goal audience of snap/flatpak
* docker. Personally I don't find docker easy (if from CLI), but some OSes (like Synology) provide easy tools for it, and people are succesfully using it.
* (too late?) a pyinstaller package for Linux? We do provide a pyinstaller package for Windows and Mac (including unrar and par2-turbo), which works great for Windows and Mac users, so why not for Linux users?

High level: more and more distro's are provding / going to provide an immutable version of their OS. So like Android. And then you can't install stuff into the OS anymore, and applications should provide everything they need in a snap / flatpak / docker package. Ubuntu Desktop 24.04 is going to have an immutable version; see https://news.itsfoss.com/ubuntu-all-snap-desktop/

About debian: I installed it (on a Raspi, so RaspberryOS) a few weeks ago, and getting rar-nonfree is still so difficult. Why? So again: debian is not great for beginners.

**Comment by sanderjo** [comment]
What I see: Linux newbie's choose some cool but not easy Linux (yes, like Debian), and then ask for help with difficult stuff. I advice them to use Ubuntu. Or docker ;-)

A bit like having no driving license/experience, and then choosing a manual gear supercar, and asking for driving instructions.

EDIT: that is often at the level that even CLI is new & scary for those users.

**Reply by jeffgeorge** [reply]
Noted, but look at the options as they stand today:
* Enjoy your out-of-date software from official repos - This sucks, and I don't see it getting any better. Like it's not even something where we can hope for backports, because right now only _sid_ has our latest released version. So it's not _one_ version behind, it's _two_. We simply move significantly faster than Debian.
* Use a different distribution - Non-starter.
* Use Docker / Snap / Flatpak / whatever - Might be ok, but again, for some weird reason all the people who have come through with this request have been _vehemently_ against Docker, and I expect they would be equally against Snap / Flatpak / etc. They seriously just want up-to-date packages for their distribution.
* Install from source - Can require a _lot_ of hand-holding to get running, especially when it comes to setting up stuff like installing a service, and I have no idea how things are going to go when it's time to update. I assume they probably won't ever update, because this might be their only service installed this way.

Is building packages for Debian and serving them up via an unofficial repo just seriously _that_ hard? If it is, cool, we'll just steer people to either Install via Docker or Install from Source, but I want to make sure building packages is truly just too hard before giving up.

**Reply by sanderjo** [reply]
> Use a different distribution - Non-starter.

Well it should be a starter. 

> Install from source - Can require a _lot_ of hand-holding to get running, 

Indeed. Unless ... there is a SABnzbd provided/included script for that. See my other post / discussion.

> especially when it comes to setting up stuff like installing a service,

Indeed. I don't help with that.

> is building packages for Debian and serving them up via an unofficial repo just seriously that hard? 

Until two years ago, jcfp's Ubuntu PPA did work for Debian. Coincidence? Not anymore, and I don't know why.

**Comment by jcfp** [comment]
Debian really isn't the distribution anyone looking for the latest releases of everything should be using. That said, I do intend to provide backports in debian, but that won't include every minor version - if only because official backports are limited to versions already accepted into unstable and testing. Version 4.1.0 is probably up next, once it has migrated to testing.

The recent increase in support questions is more than likely a side effect of the changes to how pip is handled in debian bookworm, with users now pushed towards venv and not knowing what to do now that their previous habit of polluting the system-wide python install no longer works out-of-the-box.

**Reply by Safihre** [reply]
> Debian really isn't the distribution anyone looking for the latest releases of everything should be using. That said, I do intend to provide backports in debian, but that won't include every minor version - if only because official backports are limited to versions already accepted into unstable and testing. Version 4.1.0 is probably up next, once it has migrated to testing.

Could you maybe explain how it works on Debian? We can't just "build a deb" and make our own repository somewhere like @jeffgeorge is suggesting? 
Or is there already a repository somewhere? Because you mention "backports are limited to versions already accepted into unstable and testing", so SABnzbd is updated somewhere on an official repo?

**Reply by jcfp** [reply]
In debian, packages get updated in a rolling development release ('unstable', sab 4.1.0); anything uploaded there automatically moves on to 'testing' (currently at sab 4.0.2) if it survives the CI and no release critical bugs get filed against it. Only packages present in testing at cutoff time are going to be part of the next debian release.

For every debian release, there's also official backports repositories that provide newer versions of packaged software when considered useful (sab qualifies). Now in order to ensure a clean upgrade path to the next debian release, packages cannot be added there unless they are already accepted in 'testing' - unlike a ppa where one can do anything that doesn't result in Canonical deleting it from their servers.

Setting up our own repository is of course possible as well, but a bit more work (hosting, signing, etc.) given the lack of a launchpad ppa style service for debian.

**Reply by Safihre** [reply]
Just a small Google shows me some basic info that we could do all this on Github actions and host it at Github Pages of sabnzbd.org.
For example 
https://jon.sprig.gs/blog/post/2835

Is that worth it? No clue how many debian users there are..
I can see for example that there aew about 5000 active Snap users.

**Reply by jcfp** [reply]
That would take care of the hosting and signing part, but building a deb the way they do there only works when dealing with the most basic (static, architecture independent) content. Basically, great if you want to ship a simple shell script or some background images but useless beyond that.

Even an architecture independent package such as sab has dynamically generated dependencies, integrates with and handles lots of things incuding installing and restarting system services, etc, etc. And that's before you start thinking about adding the likes of sabctools that need to be [build for every supported hardware architecture](https://buildd.debian.org/status/package.php?p=sabctools). Good luck getting github to do that for you.

**Reply by Safihre** [reply]
I see. Oke!
So why are users getting 3.7.1 now, was no 4.x release pushed to backports?

**Reply by jcfp** [reply]
Not yet, waiting for 4.1.0 to get into testing (uploaded to unstable on October 5th, should migrate to testing today).

**Reply by jeffgeorge** [reply]
Ok, so it sounds like backports is our best path forward. Soon as we get a 4.x release in backports I definitely want to sync up with you to update our Debian docs to _highly recommend_ pulling SABnzbd from backports over main.

**Reply by jcfp** [reply]
Available in bookworm-backports now ([builds of sabctools](https://buildd.debian.org/status/package.php?p=sabctools&suite=bookworm-backports) for architectures other than amd64 still pending):

```
>>> debian@debian12:~$ grep backports /etc/apt/sources.list
deb https://deb.debian.org/debian/ bookworm-backports main contrib non-free non-free-firmware
deb-src https://deb.debian.org/debian/ bookworm-backports main contrib non-free non-free-firmware

>>> debian@debian12:~$ apt policy sabnzbdplus 
sabnzbdplus:
  Installed: (none)
  Candidate: 3.7.1+dfsg-2
  Version table:
     4.1.0+dfsg-1~bpo12+1 100
        100 https://deb.debian.org/debian bookworm-backports/contrib amd64 Packages
        100 /var/lib/dpkg/status
     3.7.1+dfsg-2 500
        500 https://deb.debian.org/debian bookworm/contrib amd64 Packages

>>> debian@debian12:~$ sudo apt install sabnzbdplus/bookworm-backports
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Selected version '4.1.0+dfsg-1~bpo12+1' (Debian Backports:stable-backports [all]) for 'sabnzbdplus'
The following package was automatically installed and is no longer required:
  python3-sabyenc
Use 'sudo apt autoremove' to remove it.
The following additional packages will be installed:
  python3-sabctools
The following NEW packages will be installed:
  python3-sabctools
The following packages will be upgraded:
  sabnzbdplus
1 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 738 kB of archives.
After this operation, 194 kB of additional disk space will be used.
Do you want to continue? [Y/n] 
Get:1 https://deb.debian.org/debian bookworm-backports/main amd64 python3-sabctools amd64 7.1.2-1~bpo12+1 [65.8 kB]
Get:2 https://deb.debian.org/debian bookworm-backports/contrib amd64 sabnzbdplus all 4.1.0+dfsg-1~bpo12+1 [672 kB]
Fetched 738 kB in 0s (2,631 kB/s)
Reading changelogs... Done
Selecting previously unselected package python3-sabctools.
(Reading database ... 209440 files and directories currently installed.)
Preparing to unpack .../python3-sabctools_7.1.2-1~bpo12+1_amd64.deb ...
Unpacking python3-sabctools (7.1.2-1~bpo12+1) ...
Preparing to unpack .../sabnzbdplus_4.1.0+dfsg-1~bpo12+1_all.deb ...
Unpacking sabnzbdplus (4.1.0+dfsg-1~bpo12+1) over (3.7.1+dfsg-2) ...
Setting up python3-sabctools (7.1.2-1~bpo12+1) ...
Setting up sabnzbdplus (4.1.0+dfsg-1~bpo12+1) ...
Processing triggers for man-db (2.11.2-2) ...
Processing triggers for shared-mime-info (2.2-1) ...
Processing triggers for mailcap (3.70+nmu1) ...
Processing triggers for desktop-file-utils (0.26-1) ...
Processing triggers for gnome-menus (3.36.0-1.1) ...

>>> debian@debian12:~$ 
```

**Reply by jcfp** [reply]
I'll try to also get 4.1.0 into sloppy backports for debian 11 (bullseye), in addition to the existing backport of 3.7.1 uploaded in February.

**Comment by sanderjo** [comment]
To check sabnzbd with snap on Debian:

- I installed Debian in VM
- As root, install snapd `sudo apt install snapd`
- Reboot
- as root, install SABnzbd snap image: `snap install sabnzbd`
- Run it: `snap run sabnzbd`. Note: I need to be root because of root permissions on `/var/snap/sabnzbd/common` ... :-(  Might be a PEBKAC.
- And SABnzbd is there via http://127.0.0.1:8080 . Nice.

Easy. No difficult stuff with sources, unrar, par2, etc

![image](https://github.com/sabnzbd/sabnzbd/assets/1273502/bd7a3b5f-33af-440f-a049-b4245fdc9d6d)

**Comment by DarrenPIngram** [comment]
On 2023-10-10 10:46, Safihre wrote:
> Is that worth it? No clue how many debian users there are..
> I can see for example that there aew about 5000 active Snap users.

Personally I use Debian but I'm fine with the status quo. Others can 
have a different view/need.

I guess if it could be a simple "set up and nearly forget" for your 
team, maybe it is worth it just because, but otherwise maybe make a link 
to register interest at the download page, then you'll discover if 
sufficient interest exists [maybe for a next point version onwards]?

Darren

**Comment by DarrenPIngram** [comment]
On 2023-10-10 11:37, jcfp wrote:
> And
> that's before you start thinking about adding the likes of sabctools
> that need to be build for every supported hardware architecture [1].
> Good luck getting github to do that for you.

Maybe in that case the options might be:

a. someone writes and updates a reasonably simple guide [the Plex thing 
posted recently seemed a good-ish model from a brief glance) -- probably 
over my level but maybe such a guide could be semi-generic [e.g. install 
the latest version of x by doing y).

or

b. the users of Debian must use a package manager/Docker or if they are 
so insistent roll their own version.

The project's limited resources probably can't roll out a red carpet for 
everyone. Well, maybe if they send the developers a very large amount of 
money but you get the picture.

Darren

**Reply by jeffgeorge** [reply]
> The project's limited resources probably can't roll out a red carpet for
> everyone.

I generally agree with this statement, but I also feel like we should at least _try_ to have some reasonable options for how users can get more up-to-date versions for major distributions.

It looks like the answer for Debian is going to be:
1. If you're ok with out-of-date SABnzbd, install from apt
2. If you want to stick with apt but want newer builds, use Debian Backports
3. If you want releases as we ship them, use some kind of containerization/virtualization/isolation/packaging solution like Docker, Snap, or Flatpak.

And conveniently, Option 3 is a good general fallback solution for all distributions.

**Reply by sanderjo** [reply]
Option 4: If you're experienced with Linux commandline, you run from source, with python-venv.

Furthermore: Option 3 & 4 are valid for any Linux distro

**Comment by DarrenPIngram** [comment]
On 2023-10-10 18:18, Jeff George wrote:
> I generally agree with this statement, but I also feel like we should
> at least _try_ to have some reasonable options for how users can get
> more up-to-date versions for major distributions.

Oh I agree, but as it won't be me doing the work, I'm reluctant to 
necessarily push for something that might have limited use yet be 
possibly very time-sensitive. Naturally "more is better" when resources 
allow, even if it is a more static-type document or guide, along with 
the other options you've raised, which is still "better than nothing".

I am not privy to any statistics, should they exist, about platform and 
architecture usage [not searched, so it could be Sod's Law that they are 
immediately discoverable]. Maybe that would, if it exists, help 
determine prioritisation of resources.

And, of course, "anything" can always be built upon or improved, should 
demand and/or available resources exist.

The users should, in any case, be prepared to put a modicum of effort in 
for the privilege of using such a wonderful program :)

Darren

**Comment by thezoggy** [comment]
the other issue that many of the debian users have is after they install sab (pick your own adventure), is getting the service setup. 

many of the debian users that arrive in discord have little to no linux experience.. 
several have been raspberry pi users and are stuck using debian because of OpenMediaVault (omv). 

i've walked several through on doing the venv setup and getting things going, but then that works great for the one time until they want to update sab versions and sabctools is involved because of how strict we are with versions. 

unrelated, but the whole sabctools and version to install via os package also plagues ich777 docker version of sab as he uses apt to manage packages for his setup, so he has to manually update what gets installed (as he doesnt rely on requirements.txt).

**Reply by sanderjo** [reply]
@thezoggy  Indeed. Installing "rar-nonfree" on Raspberry OS is still difficult, isn't it?

Anyway: my answer would be ...  https://snapcraft.io/docs/installing-snap-on-raspbian . So "option 3".

**Comment by thezoggy** [comment]
> @thezoggy Indeed. Installing "rar-nonfree" on Raspberry OS is still difficult, isn't it?
> 
> Anyway: my answer would be ... https://snapcraft.io/docs/installing-snap-on-raspbian . So "option 3".

i've not heard of anyone having an issue of installing unrar on debian.

![sab-discord1](http://zoggy.net/Discord_2023-10-10_12-21-04.png)

i know ive seen fallout of users when they tried to install from apt, then install from snap. then things go to crap because you have two instances installed trying to use same stuff which then causes issues and they dont realize it when the service wont start. eventually we figure out what happen and have them just nuke one and just use one package manager 

---

looking back at discord the issues have been permission issues, getting service to work. then various things on installing.. such as:

![Discord_2023-10-10_12-17-41](https://github.com/sabnzbd/sabnzbd/assets/500882/eae24b3a-c55b-4ec5-83f1-69ea70d4d3be)

![image](https://github.com/sabnzbd/sabnzbd/assets/500882/f5e97295-a57a-4edc-841a-9b81f779dbba)
and
![image](https://github.com/sabnzbd/sabnzbd/assets/500882/ed3672d2-2d32-4e5e-867e-0e305b928650)
(which is to be expected when sabctools wont exist for them just yet)

**Reply by thezoggy** [reply]
Welp
https://9to5linux.com/raspberry-pi-os-is-now-based-on-debian-bookworm-supports-raspberry-pi-5-and-pipewire

**Comment by jeffgeorge** [comment]
Bumping this 13 month old thread since we just had another request along these lines roll through Discord.

This prompted me to follow-through on some testing I intended to do previously but never did. Now I have potential recommended installation instructions via bookworm-backports:

---

- Edit `/etc/apt/sources.list` with sudo / as root
- Confirm that your existing sources include Debian's `main`, `contrib`, and `non-free` repos. Your lines should look like:
```
deb http://deb.debian.org/debian/ bookworm main contrib non-free
deb-src http://deb.debian.org/debian/ bookworm main contrib non-free

deb http://security.debian.org/debian-security bookworm-security main contrib non-free
deb-src http://security.debian.org/debian-security bookworm-security main contrib non-free

deb http://deb.debian.org/debian/ bookworm-updates main contrib non-free
deb-src http://deb.debian.org/debian/ bookworm-updates main contrib non-free
```
- NOTE: You may have _other_ repos like `non-free-firmware`, that's fine please leave them there, just make sure you have `main`, `contrib`, and `non-free` and add whichever of them you don't already have.
- Add the `bookworm-backports` repo at the bottom:
```
deb http://deb.debian.org/debian bookworm-backports main contrib non-free
```
- Save the file
- Update apt: `sudo apt update`
	- It should show any contrib / non-free repo metadata you didn't previously have being downloaded.
- Install SABnzbd from backports: `sudo apt install -t bookworm-backports sabnzbdplus`
	- NOTE: This will JUST install the packages from backports if it provides specific versions we request (like `python3-sabctools`), but will prefer your stable packages if possible (like `unrar`).
- That's it, you've installed SABnzbd from Debian Backports.
- You can proceed to the [How to Run as a Service](https://sabnzbd.org/wiki/installation/install-ubuntu-repo#toc3) section of the Ubuntu instructions.
- Note that when you do a dist-upgrade to a new stable version of Debian you _will_ need to update your backports apt source entries along with all the others.

---

So... do these recommendations look reasonable? If they do, I'd be happy to file a PR. In the new docs, I would suggest for users to first try Docker or Flatpak, but if they _must_ install from apt, here's how to get it from Backports.

**Reply by jcfp** [reply]
The install part LGTM (see my [earlier post](https://github.com/sabnzbd/sabnzbd/discussions/2706#discussioncomment-7261373) for a real life example). 

* I don't think info on running as a service is needed, as anyone looking into using backports in all likelihood already has sab installed and setup the way they want (any existing service config will continue to work). 
* doing a dist-upgrade to a newer Debian release does **not** require an active backports repo. Unlike Ubuntu PPAs, the more "official" Debian backports impose limits that guarantee upgrade paths like so: current + current-backport --> next (_next-backport_ not required); current + current-backports-sloppy --> next + next-backports (but no need for _next-backports-sloppy_).

**Reply by jeffgeorge** [reply]
> doing a dist-upgrade to a newer Debian release does not require an active backports repo. Unlike Ubuntu PPAs, the more "official" Debian backports impose limits that guarantee upgrade paths like so: current + current-backport --> next (next-backport not required); current + current-backports-sloppy --> next + next-backports (but no need for next-backports-sloppy).

Sure, but like, they still have to update it if they want to continue receiving future updates from future-backports. Is the expectation that when Trixie launches next year (???), a trixie-backports will be made available immediately, but there just won't actually be anything in it yet? If so, I think including that note should be fine.

**Reply by jcfp** [reply]
Trixie-backports as such already exists on the mirrors [right now](https://ftp.debian.org/debian/dists/trixie-backports/). I just don't expect configuring it prior to an upgrade to the next Debian release to have any effect though, due to the default low-priority pinning for all backport repos. Same as how adding bookworm-backports onto bookworm and running a full system update does not result in anything getting pulled in from there, until one explicitly "commands" the initial install of a certain package from bookworm-backports.

**Comment by jeffgeorge** [comment]
Filed a PR to improve this here: sabnzbd/sabnzbd.github.io#343

**Reply by Safihre** [reply]
Great!
